src/huggingface_hub/hf_api.py

```diff
@@ -3381,36 +3359,26 @@ def _parse_revision_from_pr_url(pr_url: str) -> str:
     return f"refs/pr/{re_match[1]}"
 
 
-def _warn_if_truncated(
-    items: List[Any], limit: Optional[int], total_count: Optional[str]
-) -> None:
-    # TODO: remove this once pagination is properly implemented in `huggingface_hub`.
-    if total_count is None:
-        # Total count header not implemented
-        return
-
-    try:
-        total_count_int = int(total_count)
-    except ValueError:
-        # Total count header not implemented properly server-side
-        return
-
-    if len(items) == total_count_int:
-        # All items have been returned => not truncated
-        return
-
-    if limit is not None and len(items) == limit:
-        # `limit` is set => truncation is expected
-        return
-
-    # Otherwise, pagination has been enabled server-side and the output has been
-    # truncated by server => warn user.
-    warnings.warn(
-        "The list of repos returned by the server has been truncated. Listing repos"
-        " from the Hub using `list_models`, `list_datasets` and `list_spaces` now"
-        " requires pagination. To get the full list of repos, please consider upgrading"
-        " `huggingface_hub` to its latest version."
-    )
+def _paginate(path: str, params: Dict, headers: Dict) -> Iterable:
+    """Fetch a list of models/datasets/spaces and paginate through results.
+
+    For now, pagination is not mandatory on the Hub. However at some point the number of
+    repos per page will be limited for performance reasons. This helper makes `huggingface_hub`
+    compliant with future server-side updates.
+    """
+    r = requests.get(path, params=params, headers=headers)
+    hf_raise_for_status(r)
+    yield from r.json()
+
+    # If pagination is implemented server-side, follow pages
+    # Next link already contains query params
+    next_page = r.headers.get("Link")
+    while next_page is not None:
+        logger.debug(f"Pagination detected. Requesting next page: {next_page}")
+        r = requests.get(next_page, headers=headers)
+        hf_raise_for_status(r)
+        yield from r.json()
+        next_page = r.headers.get("Link")
```

Comment:
The `Link` header is of this shape:

```
Link: <http://localhost:5564/api/models?limit=10&sort=likes&direction=-1&cursor=eyIkb3IiOlt7Imxpa2VzIjoyOTQsIl9pZCI6eyIkZ3QiOiI2MjFmZmRjMTM2NDY4ZDcwOWYxNzlhMTIifX0seyJsaWtlcyI6eyIkbHQiOjI5NH19XX0%3D>; rel="next"
```

In the future it can contain other links, eg:

```
Link: <http://localhost:5564/api/models?limit=10&sort=likes&direction=-1&cursor=eyIkb3IiOlt7Imxpa2VzIjoyOTQsIl9pZCI6eyIkZ3QiOiI2MjFmZmRjMTM2NDY4ZDcwOWYxNzlhMTIifX0seyJsaWtlcyI6eyIkbHQiOjI5NH19XX0%3D>; rel="next", <http://localhost:5564/api/models?limit=10&sort=likes&direction=-1&cursor=eyIkb3IiOlt7Imxpa2VzIjoyOTQsIl9pZCI6eyIkZ3QiOiI2MjFmZmRjMTM2NDY4ZDcwOWYxNzlhMTIifX0seyJsaWtlcyI6eyIkbHQiOjI5NH19XX0%3D>; rel="previous"
```

So, basically, multiple `<$url>; rel="$description"` joined by commas. It's the same as github's APIs so you can use their response to test things: https://docs.github.com/en/rest/guides/traversing-with-pagination#link-header

Reply:
Good thanks for the heads-up !

Reply:
@coyotte508 Made the change in https://github.com/huggingface/huggingface_hub/pull/1176/commits/6f38649542db0ca0e474f374e0e075fb5517c2e3.

`requests` library already parses the `"Link"`  header which makes everything easier :)